### PR TITLE
gh-1145: Default to wrapping rng if backend is not known

### LINKED ISSUE #1145: rng backend isn't validated
### Add your issue here

# Description

When we have an optional `rng` parameter in a function, we do not currently validate its array backend. This could mean we have mixed array backends within a single function.

### PR DESCRIPTION
**Pull request by connoraird**
# Description

- Alters the existing `glass.rng.Generator` to allow it to wrap more than just NumPy RNGs.
- Adds a new RNG protocol to ensure the wrapped rng in `glass.rng.Generator` defines the required methods.
- Updates how rngs are handles within functions to ensure they always return the required array backend.

Closes: #1145 

## Changelog entry

Changed: `glass.rng.Generator` class now wraps returned values ensuring all methods return the correct array backend.

## Checks

- [x] Is your code passing linting?
- [x] Is your code passing tests?
- [x] Have you added additional tests (if required)?
- [ ] Have you modified/extended the documentation (if required)?
- [x] Have you added a one-liner changelog entry above (if required)?

### CODE REVIEW

File: tests/core/test_array_api_utils.py [hunk 1]
```diff
@@ -44,6 +44,14 @@ def test_init() -> None:
     assert isinstance(rng, glass.rng.Generator)
 
 
+@pytest.mark.skipif(not HAVE_ARRAY_API_STRICT, reason="test requires array_api_strict")
+def test_init_mix_of_backends() -> None:
+    import array_api_strict
+
+    rng = glass.rng.Generator(rng=np.random.default_rng(), xp=array_api_strict)
+    assert isinstance(rng, glass.rng.Generator)
```
Comment (paddyroddy) [inline tests/core/test_array_api_utils.py]: So to check the purpose of this PR. We are allowing a mix of RNG and `xp`? Should we be testing that more explicitly? Like here, is `rng` NumPy or `array-api-strict`? Does it matter?

File: glass/galaxies.py [hunk 2]
```diff
@@ -74,13 +74,16 @@ def _draw_nz(
         Redshifts sampled from the given source distribution.
 
     """
+    xp = array_api_compat.array_namespace(count, z, nz, use_compat=False)
```
Comment (paddyroddy) [inline glass/galaxies.py:77]: This line is just because we need an `xp` to be used within the new `Generator` class?

File: glass/galaxies.py [hunk 3]
```diff
@@ -420,8 +420,7 @@ def gaussian_phz(  # noqa: PLR0913
     sigma_0_arr = xp.asarray(sigma_0)
 
     # get default RNG if not given
```
Comment (paddyroddy) [inline glass/galaxies.py]: This comment has been left here. Can we remove it everywhere?

File: glass/rng.py [hunk 4]
```diff
@@ -36,12 +36,13 @@ def default_rng(
     if xp.__name__ == "jax.numpy":
         import glass.jax  # noqa: PLC0415
 
-        return glass.jax.Generator(seed=seed)
+        return glass.jax.Generator(seed=seed)  # ty: ignore[invalid-return-type]
```
Comment (paddyroddy) [inline glass/rng.py]: Should we be changing the output type above from `UnifiedGenerator?`

File: glass/rng.py [hunk 5]
```diff
@@ -36,12 +36,13 @@ def default_rng(
...
 
-    if xp.__name__ == "numpy":
-        return xp.random.default_rng(seed=seed)
+    import numpy as np  # noqa: PLC0415
 
-    return Generator(xp=xp, seed=seed)
+    rng = np.random.default_rng(seed=seed)
+
+    return rng if xp.__name__ == "numpy" else Generator(rng=rng, xp=xp)  # ty: ignore[invalid-return-type]
```
Comment (paddyroddy) [inline glass/rng.py]: I think what we had before is clearer, i.e.

```python
if xp.__name__ == "numpy":
    return rng

import numpy as np  # noqa: PLC0415

rng = np.random.default_rng(seed=seed)
return Generator(rng=rng, xp=xp)
```

File: glass/rng.py [hunk 6]
```diff
@@ -63,18 +64,20 @@ def rng_dispatcher(*, xp: ModuleType) -> UnifiedGenerator:
...
-    __slots__ = ("np", "rng", "xp")
+    __slots__ = ("default_dtype", "rng", "xp")
 
     def __init__(
         self,
+        *,
         xp: ModuleType,
+        rng: UnifiedGenerator | None = None,
         seed: int | IntArray = SEED,
     ) -> None:
```
Comment (paddyroddy) [inline glass/rng.py:79]: Given that they're now all keyword, how about


```suggestion
        rng: UnifiedGenerator | None = None,
        seed: int | IntArray = SEED,
        xp: ModuleType,
```

File: glass/rng.py [hunk 7]
```diff
@@ -88,11 +91,12 @@ def __init__(
...
-        import numpy  # noqa: ICN001, PLC0415
-
         self.xp = xp
-        self.np = numpy
-        self.rng = self.np.random.default_rng(seed=seed)
+        self.default_dtype = xp.float64
+        if rng is None:
+            self.rng = default_rng(seed=seed, xp=xp)
+        else:
+            self.rng = rng
```
Comment (paddyroddy) [inline glass/rng.py:104]: Whereas this, I think it's okay to do

```suggestion
        self.rng = default_rng(seed=seed, xp=xp) if rng is None else rng
```

File: glass/rng.py [hunk 8]
```diff
@@ -117,8 +121,8 @@ def random(
             Array of random floats.
 
         """
-        dtype = dtype if dtype is not None else self.np.float64
-        return self.xp.asarray(self.rng.random(size, dtype, out))  # ty: ignore[no-matching-overload]
+        dtype = dtype if dtype is not None else self.default_dtype
+        return self.xp.asarray(self.rng.random(size, out), dtype=dtype)
```
Comment (paddyroddy) [inline glass/rng.py]: I like it, but why can we do this now and why couldn't we before?

File: glass/rng.py [hunk 9]
```diff
@@ -190,8 +194,8 @@ def standard_normal(
             Array of samples from the standard normal distribution.
 
         """
-        dtype = dtype if dtype is not None else self.np.float64
-        return self.xp.asarray(self.rng.standard_normal(size, dtype, out))  # ty: ignore[no-matching-overload]
+        dtype = dtype if dtype is not None else self.default_dtype
+        return self.xp.asarray(self.rng.standard_normal(size, out), dtype=dtype)
```
Comment (paddyroddy) [inline glass/rng.py]: I like it, but why can we do this now and why couldn't we before?

File: glass/rng.py
Comment (paddyroddy) [inline glass/rng.py:81]: I'm still not overly comfortable with `Protocols`. We are duplicating the docstrings here. Is this expected behaviour? Can you find any examples which show duplication Vs not?

File: glass/points.py [hunk 10]
```diff
@@ -522,8 +521,7 @@ def positions_from_delta(  # noqa: PLR0913
     xp = array_api_compat.array_namespace(ngal, delta, bias, vis, use_compat=False)
 
     # get default RNG if not given
-    if rng is None:
-        rng = glass.rng.rng_dispatcher(xp=xp)
```
Comment (paddyroddy) [inline glass/points.py:526]: Are we still using `rng_dispatcher` at all?

**Review by paddyroddy — CHANGES_REQUESTED** [review]

File: tests/core/test_array_api_utils.py (hunk 1 above)
Comment (connoraird) [reply tests/core/test_array_api_utils.py]: Currently, we can have a mix of `rng` and `xp`  which will could in theory cause a failure because of mismatched array backends. This PR ensure that, if the `rng` provided does not match `xp`, we convert the returned `rng` arrays to the correct `xp` backend.

Here `rng` is the NumPy generator wrapped by `array-api-strict` but in theory we could have any combination of `rng` and `xp`

File: glass/galaxies.py (hunk 2 above)
Comment (connoraird) [reply glass/galaxies.py:77]: Yes

File: glass/galaxies.py (hunk 3 above)
Comment (connoraird) [reply glass/galaxies.py]: Can do, wasn't sure if it was still useful

File: glass/rng.py (hunk 4 above)
Comment (connoraird) [reply glass/rng.py]: Change it to what? An alternative is to keep UnifiedGenerator as it was and give this new protocol class a different name?

File: glass/rng.py
Comment (connoraird) [reply glass/rng.py:81]: The example I used was from cosmology api. Happy to deduplicate but I feel like it's important to have the docstrings in the protocol as that's where we are defining the signature. Is there something similar to doxygen `copydoc`?

File: glass/rng.py (hunk 8 above)
Comment (connoraird) [reply glass/rng.py]: I don't know why we couldn't before TBH.

File: glass/points.py (hunk 10 above)
Comment (connoraird) [reply glass/points.py:526]: We are using it when we know for a fact `rng` is `None` (i.e. a new instance) so we don't need to do the loop of `Generator` ->  `default_rng` -> `Generator`. Currently only in one place.

File: glass/rng.py (hunk 4 above)
Comment (connoraird) [reply glass/rng.py]: I've updated the name of the protocol

File: tests/core/test_array_api_utils.py (hunk 1 above)
Comment (paddyroddy) [reply tests/core/test_array_api_utils.py]: What is the advantage of asserting that `rng` is `glass.rng.Generator`? Is that not inevitable?

File: glass/galaxies.py (hunk 3 above)
Comment (paddyroddy) [reply glass/galaxies.py]: Well I don't mind either way, but you had removed it some places and not others

File: glass/rng.py (hunk 4 above)
Comment (paddyroddy) [reply glass/rng.py]: Will have a look, but yes the alternative was some long-winded `UnifiedGenerator | glass.jax.Generator | ...` etc.

File: tests/core/test_array_api_utils.py (hunk 1 above)
Comment (connoraird) [reply tests/core/test_array_api_utils.py]: Good point. I've updated to check the returned arrays are as expected.

**Comment by connoraird** [comment]
@paddyroddy I've made some significant changes after discovering a bug when adding the new tests with JAX.

File: glass/rng.py (hunk 8 above)
Comment (paddyroddy) [reply glass/rng.py]: This is nicer, feel the old way wasn't super readable

File: glass/points.py (hunk 10 above)
Comment (paddyroddy) [reply glass/points.py:526]: Given that `rng_dispatcher` is only `default_rng(seed=SEED, xp=xp)`, is it still worth having?

**Comment by paddyroddy** [comment]
> @paddyroddy I've made some significant changes after discovering a bug when adding the new tests with JAX.

Okay, will have a look

File: glass/rng.py
Comment (paddyroddy) [reply glass/rng.py:81]: So I did some reading and I think we should copy the one line top of the docstring in the `Protocol`, but not the rest. I've pushed this for simplicity.

File: glass/rng.py [hunk 11]
```diff
@@ -63,19 +111,23 @@ def rng_dispatcher(*, xp: ModuleType) -> UnifiedGenerator:
...
 
-    This class wraps NumPy's random number generator and returns arrays compatible
-    with the provided backend.
+    This class wraps random number generators which match the glass UnifiedGenerator
+    type and returns arrays compatible with the provided backend.
 
     """
 
-    __slots__ = ("np", "rng", "xp")
+    default_dtype: DTypeLike
```
Comment (paddyroddy) [inline glass/rng.py:67]: We could do this?

```suggestion
    __slots__ = ("default_dtype", "rng", "xp")

    default_dtype: DTypeLike
```

File: glass/rng.py [hunk 12]
```diff
@@ -88,17 +140,26 @@ def __init__(
...
-        self.np = numpy
-        self.rng = self.np.random.default_rng(seed=seed)
+        self.default_dtype = xp.float64
+
+        if rng is None:
+            if xp.__name__ == "jax.numpy":
+                import glass.jax  # noqa: PLC0415
+
+                self.rng = glass.jax.Generator(seed=seed)
+
```
Comment (paddyroddy) [inline glass/rng.py]:
```suggestion
```

File: tests/core/test_array_api_utils.py [hunk 13]
```diff
@@ -44,6 +44,31 @@ def test_init() -> None:
     assert isinstance(rng, glass.rng.Generator)
 
 
+@pytest.mark.skipif(not HAVE_ARRAY_API_STRICT, reason="test requires array_api_strict")
```
Comment (paddyroddy) [inline tests/core/test_array_api_utils.py:47]: These tests are much better

File: glass/jax.py [hunk 14]
```diff
@@ -201,6 +205,10 @@ def multinomial(
         dtype: DTypeLike = int,
     ) -> IntArray:
         """Draw samples from a multinomial distribution."""
+        # Ensure arrays are jax arrays
+        n = jnp.asarray(n)
+        pvals = jnp.asarray(pvals)
```
Comment (paddyroddy) [inline glass/jax.py:210]: Interesting that it's only these two methods that needed it

Comment (connoraird) [reply glass/jax.py:210]: In this case, it's needed because of the way we index `pvals` below `pvals[..., 0]`. I suppose `n` may not be necessary but it's a bit more complete.

File: glass/rng.py (hunk 11 above)
Comment (connoraird) [reply glass/rng.py:67]: Looks good. What's the benefit of `__slots__`?

File: glass/jax.py (hunk 14 above)
Comment (connoraird) [reply glass/jax.py:210]: Although, maybe I'm misremembering  as both instances are passing into `_shape` so that might be it.

File: glass/points.py (hunk 10 above)
Comment (connoraird) [reply glass/points.py:526]: Probably not

File: glass/rng.py (hunk 11 above)
Comment (connoraird) [reply glass/rng.py:67]: I've had a search and seems `__slots__` making it faster

**Comment by connoraird** [comment]
I've removed the Protocol class as I don't think it was actually doing anything. It definitely wasn't doing what I wanted which was restricting what RNGs could be passed to `Generator`

File: glass/jax.py (hunk 14 above)
Comment (paddyroddy) [reply glass/jax.py:210]: Remind me why this change is needed now? Is it because we are now actually testing with JAX?

File: glass/rng.py (hunk 11 above)
Comment (paddyroddy) [reply glass/rng.py:67]: Yeah, I'm not super familiar with it as a concept, but I've used it before, and it is what we had before so seems like a suitable change.

**Comment by paddyroddy** [comment]
> I've removed the Protocol class as I don't think it was actually doing anything. It definitely wasn't doing what I wanted which was restricting what RNGs could be passed to `Generator`

Okay makes sense. Would be good if @ntessore could explain them to us a bit more 😅

**Comment by paddyroddy** [comment]
Could have been a fluke, but the examples took more than 20 mins and failed. Looks like they usually take about 15 mins. Have re-run, otherwise we might have had a regression.

File: glass/rng.py [hunk 15]
```diff
@@ -88,17 +88,25 @@ def __init__(
...
+            if xp.__name__ == "jax.numpy":
+                import glass.jax  # noqa: PLC0415
+
+                self.rng = glass.jax.Generator(seed=seed)
+            else:
+                import numpy as np  # noqa: PLC0415
+
+                self.rng = np.random.default_rng(seed=seed)
+        else:
+            self.rng = rng
```
Comment (paddyroddy) [inline glass/rng.py:104]: I think this logic is nice and clear now

**Review by paddyroddy — APPROVED** [review]
Looks good, see one comment on examples timing

**Comment by paddyroddy** [comment]
Examples took 21:00
